In [2]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
model = Sequential([
    Flatten(input_shape=(28, 28)), # Ubah gambar 28x28 menjadi vektor
    Dense(128, activation='relu'), # Hidden layer 1
    Dense(64, activation='relu'),  # Hidden layer 2
    Dense(10, activation='softmax') # Output layer (10 kelas)
])

# 3. Kompilasi model
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# 4. Latih model
model.fit(X_train, y_train, epochs=5, batch_size=32, validation_data=(X_test, y_test))

# 5. Evaluasi model
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi pada data uji: {acc:.4f}")

/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.8826 - loss: 0.4040 - val_accuracy: 0.9567 - val_loss: 0.1387
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9664 - loss: 0.1087 - val_accuracy: 0.9730 - val_loss: 0.0870
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9791 - loss: 0.0673 - val_accuracy: 0.9710 - val_loss: 0.0896
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9839 - loss: 0.0483 - val_accuracy: 0.9744 - val_loss: 0.0914
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9870 - loss: 0.0403 - val_accuracy: 0.9763 - val_loss: 0.0829
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9712 - loss: 0.0989
Akurasi pada data uji: 0.9763


Coba dengan beberapa parameter lain:

* Ubah jumlah neuron di hidden layer (misal: 256 dan 128).

* Tambahkan satu hidden layer lagi.

* Bandingkan akurasi dan waktu pelatihan.

* Eksperimen dengan fungsi aktivasi Sigmoid vs ReLU.

In [3]:
import tensorflow as tf
import time
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

print("Data siap digunakan.")

Data siap digunakan.


In [4]:
print("--- Eksperimen 1: Ubah Neuron (256 & 128) ---")

# Bangun model
model_1 = Sequential([
    Flatten(input_shape=(28, 28)),     # Input layer (gambar 28x28)
    Dense(256, activation='relu'),     # Hidden layer 1 (256 neuron)
    Dense(128, activation='relu'),     # Hidden layer 2 (128 neuron)
    Dense(10, activation='softmax')    # Output layer (10 kelas)
])

# Kompilasi
model_1.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Latih model & Hitung Waktu
start_time = time.time()
history_1 = model_1.fit(X_train, y_train, epochs=5, batch_size=32, validation_split=0.2, verbose=1)
end_time = time.time()

# Evaluasi
loss_1, acc_1 = model_1.evaluate(X_test, y_test, verbose=0)
time_1 = end_time - start_time

print(f"\n[Model 1] Akurasi: {acc_1:.4f}")
print(f"[Model 1] Waktu Pelatihan: {time_1:.2f} detik")

--- Eksperimen 1: Ubah Neuron (256 & 128) ---


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 11s 7ms/step - accuracy: 0.8812 - loss: 0.4040 - val_accuracy: 0.9660 - val_loss: 0.1181
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 6ms/step - accuracy: 0.9719 - loss: 0.0900 - val_accuracy: 0.9703 - val_loss: 0.1025
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 6ms/step - accuracy: 0.9808 - loss: 0.0601 - val_accuracy: 0.9707 - val_loss: 0.0963
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9855 - loss: 0.0447 - val_accuracy: 0.9766 - val_loss: 0.0870
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 6ms/step - accuracy: 0.9893 - loss: 0.0317 - val_accuracy: 0.9732 - val_loss: 0.1004

[Model 1] Akurasi: 0.9765
[Model 1] Waktu Pelatihan: 49.93 detik


In [5]:
print("--- Eksperimen 2: Tambah Extra Hidden Layer ---")

# Bangun model (3 Hidden Layers)
model_2 = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(256, activation='relu'),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),      # <--- Hidden layer tambahan
    Dense(10, activation='softmax')
])

# Kompilasi
model_2.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Latih model & Hitung Waktu
start_time = time.time()
history_2 = model_2.fit(X_train, y_train, epochs=5, batch_size=32, validation_split=0.2, verbose=1)
end_time = time.time()

# Evaluasi
loss_2, acc_2 = model_2.evaluate(X_test, y_test, verbose=0)
time_2 = end_time - start_time

print(f"\n[Model 2] Akurasi: {acc_2:.4f}")
print(f"[Model 2] Waktu Pelatihan: {time_2:.2f} detik")

--- Eksperimen 2: Tambah Extra Hidden Layer ---
Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 11s 7ms/step - accuracy: 0.8786 - loss: 0.4137 - val_accuracy: 0.9541 - val_loss: 0.1560
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9669 - loss: 0.1030 - val_accuracy: 0.9656 - val_loss: 0.1143
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 6ms/step - accuracy: 0.9789 - loss: 0.0674 - val_accuracy: 0.9707 - val_loss: 0.1022
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 6ms/step - accuracy: 0.9857 - loss: 0.0488 - val_accuracy: 0.9710 - val_loss: 0.1009
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9878 - loss: 0.0369 - val_accuracy: 0.9754 - val_loss: 0.0855

[Model 2] Akurasi: 0.9771
[Model 2] Waktu Pelatihan: 48.86 detik


In [6]:
print("--- Eksperimen 3: Ganti Aktivasi ke Sigmoid ---")

# Bangun model dengan Sigmoid
model_sigmoid = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(256, activation='sigmoid'),  # Ganti ReLU ke Sigmoid
    Dense(128, activation='sigmoid'),  # Ganti ReLU ke Sigmoid
    Dense(10, activation='softmax')
])

# Kompilasi
model_sigmoid.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Latih model & Hitung Waktu
start_time = time.time()
history_sig = model_sigmoid.fit(X_train, y_train, epochs=5, batch_size=32, validation_split=0.2, verbose=1)
end_time = time.time()

# Evaluasi
loss_sig, acc_sig = model_sigmoid.evaluate(X_test, y_test, verbose=0)
time_sig = end_time - start_time

print(f"\n[Model Sigmoid] Akurasi: {acc_sig:.4f}")
print(f"[Model Sigmoid] Waktu Pelatihan: {time_sig:.2f} detik")

--- Eksperimen 3: Ganti Aktivasi ke Sigmoid ---
Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 11s 7ms/step - accuracy: 0.7906 - loss: 0.7793 - val_accuracy: 0.9342 - val_loss: 0.2168
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 7ms/step - accuracy: 0.9433 - loss: 0.1896 - val_accuracy: 0.9559 - val_loss: 0.1500
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 6ms/step - accuracy: 0.9604 - loss: 0.1325 - val_accuracy: 0.9669 - val_loss: 0.1154
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 6ms/step - accuracy: 0.9726 - loss: 0.0900 - val_accuracy: 0.9673 - val_loss: 0.1030
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9795 - loss: 0.0691 - val_accuracy: 0.9737 - val_loss: 0.0883

[Model Sigmoid] Akurasi: 0.9742
[Model Sigmoid] Waktu Pelatihan: 49.70 detik


## 1. Ringkasan Hasil
| Model | Konfigurasi | Fungsi Aktivasi | Waktu Training (detik) | Akurasi Test |
| :--- | :--- | :--- | :--- | :--- |
| **Model 1** | 2 Hidden Layers (256, 128) | ReLU | **49.93 s** | **0.9765 (97.65%)** |
| **Model 2** | 3 Hidden Layers (256, 128, 64) | ReLU | **48.86 s** | **0.9771 (97.71%)** |
| **Model 3** | 2 Hidden Layers (256, 128) | Sigmoid | **49.70 s** | **0.9742 (97.42%)** |

---

## 2. Analisis Eksperimen

### A. Pengaruh Jumlah Neuron (Model 1)
Dengan menggunakan konfigurasi neuron yang cukup besar (256 dan 128), model mampu mencapai akurasi yang sangat baik (**97.65%**) hanya dalam 5 epoch.
* **Observasi:** Kapasitas model sudah cukup mumpuni untuk menangkap pola tulisan tangan pada dataset MNIST.

### B. Pengaruh Penambahan Layer (Model 2)
Menambahkan satu hidden layer ekstra (64 neuron) memberikan hasil akurasi tertinggi yaitu **97.71%**.
* **Akurasi:** Terjadi sedikit peningkatan akurasi dibandingkan Model 1 (+0.06%), menunjukkan bahwa jaringan yang lebih dalam (*deeper network*) sedikit membantu dalam membedakan fitur yang lebih spesifik.
* **Waktu:** Secara unik, waktu pelatihan tercatat sedikit lebih cepat (**48.86 detik**) dibandingkan Model 1. Hal ini kemungkinan besar disebabkan oleh variabilitas beban server Google Colab saat *runtime*, karena secara teoritis penambahan layer seharusnya menambah beban komputasi. Namun, perbedaannya sangat tipis (< 1 detik).

### C. Perbandingan Aktivasi: ReLU vs Sigmoid (Model 3)
Mengganti fungsi aktivasi menjadi Sigmoid menghasilkan akurasi terendah di antara ketiga percobaan (**97.42%**).
* **Analisis:** Meskipun selisihnya tidak terlalu jauh, hasil ini mengonfirmasi bahwa **ReLU** bekerja lebih baik untuk *hidden layer* pada jaringan saraf modern dibandingkan Sigmoid. ReLU membantu model belajar lebih cepat dan menghindari masalah *vanishing gradient* yang membuat Sigmoid kurang efisien dalam mencapai loss minimal.

---

## 3. Kesimpulan Akhir
Berdasarkan data percobaan di atas:
1. **Konfigurasi Terbaik:** **Model 2** (3 Hidden Layers dengan ReLU) adalah yang terbaik karena menghasilkan akurasi tertinggi (97.71%).
2. **Fungsi Aktivasi:** **ReLU** terbukti lebih unggul dibandingkan Sigmoid untuk kasus ini.
3. **Efisiensi:** Penambahan layer meningkatkan akurasi tanpa mengorbankan waktu pelatihan secara signifikan pada skala dataset ini.